tlvc/init/main.c

```diff
@@ -352,19 +363,23 @@ static void INITPROC comirq(char *line)
 		if (irq[i]) set_serial_irq(i, irq[i]);
 }
 
-static void INITPROC init_chs(int dev, char *line)
+static void INITPROC init_hdparms(int dev, char *line)
 {
-	int i, val[3];
+	int i, val[CHS_ARR_SIZE] = {0};
```

Comment:
Hello! A couple of comments: initializing auto (stack) variables like this is very expensive. Under the hood, `gcc-ia16` will generate code to call `memset` (which is normally found in libgcc.a but is overridden in this case by the same named routine in lib/string.S):
```
   9:	b8 06 00             	mov    $0x6,%ax
   c:	50                   	push   %ax
   d:	b8 00 00             	mov    $0x0,%ax
  10:	50                   	push   %ax
  11:	52                   	push   %dx
  12:	16                   	push   %ss
  13:	1f                   	pop    %ds
  14:	e8 fe ff             	call   15 <init_hdparms+0x15>
			15: R_386_PC16	memset
  17:	83 c4 06             	add    $0x6,%sp
```

It might be better to consider handling zero initialization as part of an else clause, as in kernel code calling implicit procedures can have unintended side effects. FYI.

Reply:
I just noticed in the code above that `gcc-ia16` actually resets DS from SS to guarantee that `memset` has the correctly addressable segment for the variable to be cleared. This is a good example of unintended consequences, as if the kernel for some reason at the moment required DS != SS, this would reset DS without one knowing it, all from a relatively innocuous variable declaration!

Reply:
Wow, thanks @ghaerr - this is doubly useful. First the knowledge (actually somewhat scary about the DS reset), secondly the fact that I forgot to delete the entire thing. The `init_hdparm()` function is no longer used.

Appreciated.

Reply:
Further analysis shows that the push SS/pop DS is eliminated with optimization on, which is the default for the kernel compilation (`memset` is still used to initialize the array variable). Initializations of single `char` or `int` variables are performed with `mov` instructions. 

Interestingly, an initialization like `int val[3] = {1};` first does a `memset(...,0,6)` for the entire array, followed by a `mov $1,-6(%bp)` to the first element, and `int val[3] = {1, 1};` does a `memset` followed by two moves to the first two elements!!

Finally, a full (non-zero) initialization like `int val[3] = {1, 1, 1};` generates three words in the data segment of value 1, and then uses internal `memcpy` to initialize the auto stack variable from the data segment. That I didn't know - and I think I'll stay away from any auto array initializations knowing that!

Reply:
Again, very interesting indeed @ghaerr. Incidentally, while working on this PR, I became unsure about how array initialization worked, wondering about the (possible) difference between say `int i[10] = {0,}` and `int i[10] = {0}`.

What I found was both educating and surprising - including the fact that this is all gcc only and thus non-standard. OTOH - to most of us, gcc IS the standard and this is what counts.

Anyway, this is the way I now understand array initialization:
- There is no difference between the two variants above.
- Zero has special significance as an initializer. 
- While `int i[10] = {0};` does what you'd expect and initializes all elements to 0, `int[10] = {1};` does not what I'd expected. It sets the first element to 1, the rest to zero.

There is more to it, but this is the gist of it - and it seems coherent with the discoveries (initialization mechanisms) you've made.

This knowledge - and the real cost of array initialization as you've demonstrated - will surely change my focus on initializers and initializing. There may be many bugs out there caused by erroneous assumptions about how initializers actually work. 

Thanks.

Reply:
> While int i[10] = {0}; does what you'd expect and initializes all elements to 0

Yes, that is standardized in C89.

> int[10] = {1}; does not what I'd expected. It sets the first element to 1, the rest to zero.

Yes, that also is standardized in C89 - "If there are fewer initializers in a list than there are members of an aggregate, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration." In other words, non-specified values in an initialization list will be set to 0.

> the real cost of array initialization as you've demonstrated - will surely change my focus on initializers and initializing.

Good, I'm glad I was able to help. What I try to do is guess at what I would think the compiler is generating for any line of C written... and then occasionally verify because sometimes it isn't what one thinks! GCC has all sorts of other "optimizations" one has to watch out for; one that's gotten me a number of times is `printf("%s\n", str)` is converted to `puts(str)` without any mention. `memset` and `memcpy` are called behind the scenes for lots of things, including initialization and structure assignment.

Reply:
So C89 it is, thanks for pointing that out - I'm guilty at putting too much trust in the source I found on the net. 
It seems even after all these years, there's still stuff to be learned about C.